core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/StrictEvaluationStrategy.java

```diff
@@ -248,31 +249,38 @@ public TupleExpr optimize(TupleExpr expr, EvaluationStatistics evaluationStatist
 	@Override
 	public CloseableIteration<BindingSet, QueryEvaluationException> evaluate(TupleExpr expr, BindingSet bindings)
 			throws QueryEvaluationException {
+
+		CloseableIteration<BindingSet, QueryEvaluationException> ret;
+
 		if (expr instanceof StatementPattern) {
-			return evaluate((StatementPattern) expr, bindings);
+			ret = evaluate((StatementPattern) expr, bindings);
 		} else if (expr instanceof UnaryTupleOperator) {
-			return evaluate((UnaryTupleOperator) expr, bindings);
+			ret = evaluate((UnaryTupleOperator) expr, bindings);
 		} else if (expr instanceof BinaryTupleOperator) {
-			return evaluate((BinaryTupleOperator) expr, bindings);
+			ret = evaluate((BinaryTupleOperator) expr, bindings);
 		} else if (expr instanceof SingletonSet) {
-			return evaluate((SingletonSet) expr, bindings);
+			ret = evaluate((SingletonSet) expr, bindings);
 		} else if (expr instanceof EmptySet) {
-			return evaluate((EmptySet) expr, bindings);
+			ret = evaluate((EmptySet) expr, bindings);
 		} else if (expr instanceof ExternalSet) {
-			return evaluate((ExternalSet) expr, bindings);
+			ret = evaluate((ExternalSet) expr, bindings);
 		} else if (expr instanceof ZeroLengthPath) {
-			return evaluate((ZeroLengthPath) expr, bindings);
+			ret = evaluate((ZeroLengthPath) expr, bindings);
 		} else if (expr instanceof ArbitraryLengthPath) {
-			return evaluate((ArbitraryLengthPath) expr, bindings);
+			ret = evaluate((ArbitraryLengthPath) expr, bindings);
 		} else if (expr instanceof BindingSetAssignment) {
-			return evaluate((BindingSetAssignment) expr, bindings);
+			ret = evaluate((BindingSetAssignment) expr, bindings);
 		} else if (expr instanceof TripleRef) {
-			return evaluate((TripleRef) expr, bindings);
+			ret = evaluate((TripleRef) expr, bindings);
 		} else if (expr == null) {
 			throw new IllegalArgumentException("expr must not be null");
 		} else {
 			throw new QueryEvaluationException("Unsupported tuple expr type: " + expr.getClass());
 		}
+
+		// set resultsSizeActual to at least be 0 so we can track iterations that don't procude anything
+		expr.setResultSizeActual(Math.max(0, expr.getResultSizeActual()));
+		return new ResultSizeCountingIterator(ret, expr);
 	}
```

Comment:
@jeenbroekstra this is what I am proposing now. A general way to track result sizes across all iterators in the query plan. This also makes it simpler to enable or disable. I was thinking of something along the lines of:
```
if(trackResultSize()){
    expr.setResultSizeActual(Math.max(0, expr.getResultSizeActual()));
   ret = new ResultSizeCountingIterator(ret, expr);
}

return ret;
```

How can we make this configurable for the user?

Reply:
ping @jeenbroekstra

Reply:
Sorry, missed this earlier, and you caught me at the end of my day. I'll take a closer look tomorrow.

Reply:
I like it, makes sense. As for making this configurable: for now I'd just make it an optional additional constructor argument on StrictEvaluationStrategy (similar to, for example, the `iterationCacheSyncTreshold` configurable setting). How we when expose this to users to tweak is a separate discussion I think: it could take form of a setting on individual Query objects perhaps, which would be more flexible than a per-repository setting.